File: backend/app/services/php.py
from pathlib import Path

from app.core.config import settings
from app.schemas.schemas import PhpConfigUpdate
from app.services.shell import shell

SUPPORTED_PHP_VERSIONS = ("5.6", "7.4", "8.0", "8.1", "8.2", "8.3", "8.4", "8.5")


def _safe_ini_value(value: str) -> str:
    if "\n" in value or "\r" in value or "\x00" in value:
        raise ValueError("Invalid PHP ini value")
    return value


def update_php_ini(payload: PhpConfigUpdate) -> str:
    php_version = payload.php_version
    if php_version not in SUPPORTED_PHP_VERSIONS:
        allowed = ", ".join(sorted(SUPPORTED_PHP_VERSIONS))
        raise ValueError(f"Unsupported PHP version. Allowed: {allowed}")
    display_errors = "On" if str(payload.display_errors).lower() in {"1", "true", "on", "yes"} else "Off"
    content = "\n".join([
        f"display_errors = {display_errors}",
        f"memory_limit = {_safe_ini_value(payload.memory_limit)}",
        f"upload_max_filesize = {_safe_ini_value(payload.upload_max_filesize)}",
        f"post_max_size = {_safe_ini_value(payload.post_max_size)}",
        f"max_execution_time = {int(payload.max_execution_time)}",
        f"max_input_time = {int(payload.max_input_time)}",
        f"max_input_vars = {int(payload.max_input_vars)}",
        "",
    ])
    target = Path(f"/etc/php/{php_version}/fpm/conf.d/99-snpanel.ini")
    if settings.command_dry_run:
        return content
    shell.privileged(
        "php-config-write",
        helper_args=[php_version],
        input=content,
        fallback=[
            "bash",
            "-lc",
            "cat > /etc/php/$1/fpm/conf.d/99-snpanel.ini && systemctl restart php$1-fpm",
            "snpanel-php-config-write",
            php_version,
        ],
    )
    return str(target)


PHP_CONFIG_KEYS = {
    "display_errors": "Off",
    "memory_limit": "1024M",
    "upload_max_filesize": "1024M",
    "post_max_size": "1024M",
    "max_execution_time": "300",
    "max_input_time": "600",
    "max_input_vars": "10000",
}


def default_php_config(php_version: str) -> dict:
    if php_version not in SUPPORTED_PHP_VERSIONS:
        allowed = ", ".join(sorted(SUPPORTED_PHP_VERSIONS))
        raise ValueError(f"Unsupported PHP version. Allowed: {allowed}")
    return {
        "php_version": php_version,
        "display_errors": PHP_CONFIG_KEYS["display_errors"],
        "memory_limit": PHP_CONFIG_KEYS["memory_limit"],
        "upload_max_filesize": PHP_CONFIG_KEYS["upload_max_filesize"],
        "post_max_size": PHP_CONFIG_KEYS["post_max_size"],
        "max_execution_time": int(PHP_CONFIG_KEYS["max_execution_time"]),
        "max_input_time": int(PHP_CONFIG_KEYS["max_input_time"]),
        "max_input_vars": int(PHP_CONFIG_KEYS["max_input_vars"]),
    }


def restore_default_php_ini(php_version: str) -> str:
    return update_php_ini(PhpConfigUpdate(**default_php_config(php_version)))


def read_php_ini(php_version: str | None = None) -> dict:
    # Resolved against what this machine actually has. SUPPORTED_PHP_VERSIONS is
    # the set the panel knows how to read, not the set that is installed, so a
    # supported-but-absent version passed the check below and then produced an
    # error naming /etc/php/<version>/fpm - a path the operator never chose.
    from app.services.php_tune import resolve_version

    php_version = resolve_version(php_version)
    if php_version not in SUPPORTED_PHP_VERSIONS:
        allowed = ", ".join(sorted(SUPPORTED_PHP_VERSIONS))
        raise ValueError(f"Unsupported PHP version. Allowed: {allowed}")
    values = dict(PHP_CONFIG_KEYS)
    for path in [
        Path(f"/etc/php/{php_version}/fpm/php.ini"),
        Path(f"/etc/php/{php_version}/fpm/conf.d/99-snpanel.ini"),
    ]:
        if not path.exists():
            continue
        for line in path.read_text(encoding="utf-8", errors="ignore").splitlines():
            line = line.strip()
            if not line or line.startswith(";") or "=" not in line:
                continue
            key, value = [part.strip() for part in line.split("=", 1)]
            if key in values:
                values[key] = value
    values["php_version"] = php_version
    values["max_execution_time"] = int(values["max_execution_time"])
    values["max_input_time"] = int(values["max_input_time"])
    values["max_input_vars"] = int(values["max_input_vars"])
    return values


def list_installed_php() -> list[str]:
    """List PHP versions that are currently installed on the system."""
    installed = []
    for version in SUPPORTED_PHP_VERSIONS:
        fpm_path = Path(f"/etc/php/{version}/fpm/php-fpm.conf")
        if fpm_path.exists():
            installed.append(version)
    return sorted(installed, key=lambda v: [int(x) for x in v.split(".")])


def install_php(php_version: str) -> dict:
    """Install a PHP version or repair the SNPanel extension set via apt."""
    if php_version not in SUPPORTED_PHP_VERSIONS:
        allowed = ", ".join(sorted(SUPPORTED_PHP_VERSIONS))
        raise ValueError(f"Unsupported PHP version. Allowed: {allowed}")

    already_installed = Path(f"/etc/php/{php_version}/fpm/php-fpm.conf").exists()

    if settings.command_dry_run:
        action = "repair" if already_installed else "install"
        return {"status": "dry_run", "message": f"Would {action} php{php_version} and SNPanel extensions"}

    result = shell.privileged(
        "php-install",
        helper_args=[php_version],
        fallback=[
            "apt-get",
            "install",
            "-y",
            f"php{php_version}-fpm",
            f"php{php_version}-cli",
            f"php{php_version}-mysql",
            f"php{php_version}-sqlite3",
            f"php{php_version}-curl",
            f"php{php_version}-gd",
            f"php{php_version}-mbstring",
            f"php{php_version}-xml",
            f"php{php_version}-zip",
            f"php{php_version}-opcache",
            f"php{php_version}-intl",
            f"php{php_version}-bcmath",
            f"php{php_version}-redis",
            f"php{php_version}-imagick",
        ],
    )
    return {"status": "ensured" if already_installed else "installed", "version": php_version, "output": result.stdout}

File: crates/snpanel-api/src/php.rs
//! PHP versions and the panel's `99-snpanel.ini`.
//!
//! Source: `app.services.php`, plus the two functions from `php_tune` and
//! `system` that decide *which* version a request without one acts on.
//!
//! That decision is the part worth reading. `settings.default_php_version` is
//! a preference, not a fact: on a machine that does not have that version,
//! acting on it hands the privileged helper a path that does not exist and
//! shows the operator "Action failed" naming a directory they never chose.

use std::path::{Path, PathBuf};

/// Source: `SUPPORTED_PHP_VERSIONS` - what the panel knows how to read, which
/// is not the same as what is installed.
pub const SUPPORTED_PHP_VERSIONS: &[&str] =
    &["5.6", "7.4", "8.0", "8.1", "8.2", "8.3", "8.4", "8.5"];

/// Source: `PHP_CONFIG_KEYS`.
pub const DEFAULTS: &[(&str, &str)] = &[
    ("display_errors", "Off"),
    ("memory_limit", "1024M"),
    ("upload_max_filesize", "1024M"),
    ("post_max_size", "1024M"),
    ("max_execution_time", "300"),
    ("max_input_time", "600"),
    ("max_input_vars", "10000"),
];

fn allowed_list() -> String {
    let mut all: Vec<&str> = SUPPORTED_PHP_VERSIONS.to_vec();
    all.sort_unstable();
    all.join(", ")
}

/// Source: `installed_php_versions` - oldest first.
pub fn installed_php_versions() -> Vec<String> {
    let Ok(entries) = std::fs::read_dir("/etc/php") else {
        return Vec::new();
    };
    let mut found: Vec<String> = entries
        .filter_map(Result::ok)
        .filter_map(|e| {
            let version = e.file_name().to_string_lossy().into_owned();
            e.path()
                .join("fpm/php-fpm.conf")
                .exists()
                .then_some(version)
        })
        .collect();
    found.sort();
    found.dedup();
    found.sort_by_key(|v| version_key(v));
    found
}

/// `[int(x) for x in v.split(".")]`, with anything unparseable last.
fn version_key(version: &str) -> (u32, u32) {
    let mut parts = version.split('.');
    let major = parts
        .next()
        .and_then(|p| p.parse().ok())
        .unwrap_or(u32::MAX);
    let minor = parts.next().and_then(|p| p.parse().ok()).unwrap_or(0);
    (major, minor)
}

/// Source: `system.default_php_version`.
///
/// The configured preference wins when the machine has it - an operator who
/// runs several versions and names one means it. Otherwise the newest
/// installed, and only then the preference as a last resort.
pub fn default_php_version(configured: &str) -> String {
    let installed = installed_php_versions();
    let configured = configured.trim();
    if !configured.is_empty() && installed.iter().any(|v| v == configured) {
        return configured.to_string();
    }
    if let Some(last) = installed.last() {
        return last.clone();
    }
    if configured.is_empty() {
        "8.4".to_string()
    } else {
        configured.to_string()
    }
}

/// Source: `php_tune.resolve_version`.
pub fn resolve_version(requested: Option<&str>, configured: &str) -> Result<String, String> {
    let requested = requested.unwrap_or("").trim();
    if requested.is_empty() {
        return Ok(default_php_version(configured));
    }
    if installed_php_versions().iter().any(|v| v == requested) {
        return Ok(requested.to_string());
    }
    if !SUPPORTED_PHP_VERSIONS.contains(&requested) {
        // Nonsense, rather than a stale default. Refusing is right: nobody
        // meant this, and silently tuning some other version would hide it.
        return Err(format!(
            "Unsupported PHP version: {requested}. Allowed: {}",
            allowed_list()
        ));
    }
    // A version the panel supports and this machine does not have: the
    // interface's stale default arriving, not a choice anyone made.
    Ok(default_php_version(configured))
}

fn ini_path(version: &str, name: &str) -> PathBuf {
    Path::new("/etc/php").join(version).join("fpm").join(name)
}

/// Source: `read_php_ini` - the defaults, then `php.ini`, then the panel's
/// own file, so the panel's value wins where it sets one.
pub fn read_php_ini(
    requested: Option<&str>,
    configured: &str,
) -> Result<serde_json::Value, String> {
    let version = resolve_version(requested, configured)?;
    if !SUPPORTED_PHP_VERSIONS.contains(&version.as_str()) {
        return Err(format!(
            "Unsupported PHP version. Allowed: {}",
            allowed_list()
        ));
    }
    let mut values: Vec<(String, String)> = DEFAULTS
        .iter()
        .map(|(k, v)| (k.to_string(), v.to_string()))
        .collect();
    for path in [
        ini_path(&version, "php.ini"),
        ini_path(&version, "conf.d/99-snpanel.ini"),
    ] {
        let Ok(text) = std::fs::read_to_string(&path) else {
            continue;
        };
        for line in text.lines() {
            let line = line.trim();
            if line.is_empty() || line.starts_with(';') {
                continue;
            }
            let Some((key, value)) = line.split_once('=') else {
                continue;
            };
            let (key, value) = (key.trim(), value.trim());
            if let Some(slot) = values.iter_mut().find(|(k, _)| k == key) {
                slot.1 = value.to_string();
            }
        }
    }

    let get = |name: &str| -> String {
        values
            .iter()
            .find(|(k, _)| k == name)
            .map(|(_, v)| v.clone())
            .unwrap_or_default()
    };
    // The three numeric fields come back as numbers, the way the Python casts
    // them; a value the file carries that is not a number falls back to the
    // default rather than failing the whole read.
    let number = |name: &str| -> i64 {
        let default = DEFAULTS
            .iter()
            .find(|(k, _)| *k == name)
            .map(|(_, v)| *v)
            .unwrap_or("0");
        get(name)
            .parse::<i64>()
            .unwrap_or_else(|_| default.parse().unwrap_or(0))
    };
    Ok(serde_json::json!({
        "php_version": version,
        "display_errors": get("display_errors"),
        "memory_limit": get("memory_limit"),
        "upload_max_filesize": get("upload_max_filesize"),
        "post_max_size": get("post_max_size"),
        "max_execution_time": number("max_execution_time"),
        "max_input_time": number("max_input_time"),
        "max_input_vars": number("max_input_vars"),
    }))
}

/// Source: `_safe_ini_value` - a newline would let one setting become two.
pub fn safe_ini_value(value: &str) -> Result<&str, String> {
    if value.contains('\n') || value.contains('\r') || value.contains('\0') {
        return Err("Invalid PHP ini value".to_string());
    }
    Ok(value)
}

/// Source: `default_php_config`.
pub fn default_php_config(version: &str) -> Result<serde_json::Value, String> {
    if !SUPPORTED_PHP_VERSIONS.contains(&version) {
        return Err(format!(
            "Unsupported PHP version. Allowed: {}",
            allowed_list()
        ));
    }
    Ok(serde_json::json!({
        "php_version": version,
        "display_errors": "Off",
        "memory_limit": "1024M",
        "upload_max_filesize": "1024M",
        "post_max_size": "1024M",
        "max_execution_time": 300,
        "max_input_time": 600,
        "max_input_vars": 10000,
    }))
}

/// Source: the body `update_php_ini` writes.
pub struct IniUpdate {
    pub display_errors: String,
    pub memory_limit: String,
    pub upload_max_filesize: String,
    pub post_max_size: String,
    pub max_execution_time: i64,
    pub max_input_time: i64,
    pub max_input_vars: i64,
}

impl IniUpdate {
    pub fn render(&self) -> Result<String, String> {
        // `"On" if str(display_errors).lower() in {"1","true","on","yes"}`
        let display = match self.display_errors.to_lowercase().as_str() {
            "1" | "true" | "on" | "yes" => "On",
            _ => "Off",
        };
        Ok(format!(
            "display_errors = {display}\n\
             memory_limit = {}\n\
             upload_max_filesize = {}\n\
             post_max_size = {}\n\
             max_execution_time = {}\n\
             max_input_time = {}\n\
             max_input_vars = {}\n",
            safe_ini_value(&self.memory_limit)?,
            safe_ini_value(&self.upload_max_filesize)?,
            safe_ini_value(&self.post_max_size)?,
            self.max_execution_time,
            self.max_input_time,
            self.max_input_vars,
        ))
    }
}

/// Where `php-config-write` puts it.
pub fn config_target(version: &str) -> String {
    format!("/etc/php/{version}/fpm/conf.d/99-snpanel.ini")
}

/// Source: `list_installed_php`.
pub fn list_installed_php() -> Vec<String> {
    let mut found: Vec<String> = SUPPORTED_PHP_VERSIONS
        .iter()
        .filter(|v| Path::new(&format!("/etc/php/{v}/fpm/php-fpm.conf")).exists())
        .map(|v| v.to_string())
        .collect();
    found.sort_by_key(|v| version_key(v));
    found
}

/// The apt package set `install_php` falls back to.
pub fn install_packages(version: &str) -> Vec<String> {
    [
        "fpm", "cli", "mysql", "sqlite3", "curl", "gd", "mbstring", "xml", "zip", "opcache",
        "intl", "bcmath", "redis", "imagick",
    ]
    .iter()
    .map(|part| format!("php{version}-{part}"))
    .collect()
}
